### PR TITLE
Clean up ffi coverage file, add to release checklist

### PR DESCRIPTION
**Pull request by Manishearth**
We *really* should keep `missing_apis.txt` small or empty, and we should also ensure missing APIs block releases.

I've emptied the file, moving all of its pieces to different sections of the allowlist. I've also repopulated the "stuff we do plan to expose for 1.0" as a "stuff we plan to expose next release" section, and plan to write the FFI for everything there.

I've added this to the release checklist as well (aside: we should have a "pre-release checklist" that contains issues like "missing FFI" that are not immediate fixes. The release process page mentions that there will be an issue with a checklist, but does not indicate anything as to its contents.)

### CODE REVIEW

**Review by sffc — CHANGES_REQUESTED** [review]
I see missing_apis.txt as containing a list of tokens to be triaged. Some of them can be triaged to the ignorelist, but others should be added rather than ignorelisted.

In other words, I'm okay with the entries that are added to the ignorelist with the intent that it is the correct place to save them, but I don't think we should be using the ignorelist to temporarily store tokens that need to get API.

**Comment by sffc** [comment]
I don't understand what's wrong with the file not being empty and having it be a release task to triage the entries (something we could have done in 1.1 but didn't do).

Ideally, people will add FFIs in the PRs where they add the Rust public functions. This doesn't always happen for the same reason we don't always have tests, docs, and benches. Perhaps in the post-1.0 world we should get more strict about blocking PRs until all these requirements are met.

In general, I am not very happy with the fact we're using a list of strings in a .rs file to perform our ignorelisting. I don't want to make that _even more_ front and center to the process.

If you want to keep the file empty, then just delete the file and make the CI job fail if verify-coverage produces a nonempty result.

**Comment by Manishearth** [comment]
> I don't understand what's wrong with the file not being empty

Because by the time release rolls around it's usually too late. I do not think people are consistent about that, and adding a line to an empty file raises more alarm bells than adding a line to a file full of junk.

--------

Ultimately, though, I do not feel strongly about keeping it empty. What I feel strongly about is that I think we need to have pressure towards triaging it _early_. If we can achieve that with a more structured _pre_ release checklist (which I want to propose anyway) I'm happy with that. I'm curious to hear what you think about process changes that will ensure these get triaged and handled in a timely manner.



Here is my concrete proposal for a comment on that file:

> This file contains all APIs that are exposed in Rust but not over FFI, and not allowlisted in IGNORED_PATHS in tools/ffi_coverage/main.rs
> 
> If adding an API that expands this file, please consider:
> 
>  - Whether you can expose that API over FFI in the same PR (and if you do so, add a `rust_link` annotation)
>  - Whether the API already has its functionality exposed over FFI (if so, add a potentially-`hidden` `rust_link` annotation to that API)
>  - Whether that API is rust-specific functionality that need not be exposed (if so, add it to the allowlist, or add a `hidden` `rust_link` annotation to a related API)
>  - Whether that API should be punted for later in FFI (if so, please check in with [usernames])
>  - Whether the API is experimental (if so, add it to the allowlist as experimental)
> 
> It is acceptable to temporarily have APIs in this file that you plan to add in a soon-upcoming PR.
>
> Please check in with [usernames] if you have questions



So one of the things I'm also hoping to do is eliminating a lot of point #3 as well: in the case of rust-specific functionality we should always use `hidden` `rust_link` on the related API.

I want to go on a crusade to clean out the allowlist, probably putting it in a separate file and organizing it more.

**Comment by Manishearth** [comment]
> I don't understand what's wrong with the file not being empty

Also the other problem I have with this is that the file is _large_, it's not easy to see what's going on there. With the main.rs allowlist we can write a small number of items so it's immediately clear what's going on. With this change for example it's quite clear: the things we need for the next release are exemplar, uniset, names, and ranges APIs.

**Comment by Manishearth** [comment]
> Also the other problem I have with this is that the file is _large_, it's not easy to see what's going on there. With the main.rs allowlist we can write a small number of items so it's immediately clear what's going on. With this change for example it's quite clear: the things we need for the next release are exemplar, uniset, names, and ranges APIs.

Actually, perhaps this is easy to resolve: if you come across a struct that is not exposed and not allowlisted, do not even bother checking methods.

**Comment by Manishearth** [comment]
I've reverted the allowlisting of todos. PTAL.

File: ffi/diplomat/tests/missing_apis.txt [hunk 1]
```diff
@@ -1,51 +1,22 @@
...
+#
+#  - Whether you can expose that API over FFI in the same PR (and if you do so, be sure to add a `rust_link` annotation)
+#  - Whether the API already has its functionality exposed over FFI (if so, add a potentially-`hidden` `rust_link` annotation to the corresponding FFI API)
+#  - Whether that API is rust-specific functionality that need not be exposed (if so, add it to the allowlist with a note, or add a `hidden` `rust_link` annotation to a related API if possible)
+#  - Whether that API should be punted for later in FFI (if so, please check in with @Manishearth, @robertbastian, or @sffc)
+#  - Whether the API is experimental (if so, add it to the allowlist as experimental)
+#
+# It is acceptable to temporarily have APIs in this file that you plan to add in a soon-upcoming PR.
+#
+# Please check in with @Manishearth, @robertbastian, or @sffc if you have questions
```
Comment (sffc) [inline ffi/diplomat/tests/missing_apis.txt:14]: Nit (optional): Please use a group like `@unicode-org/icu4x-owners` rather than individual names

**Review by sffc — APPROVED** [review]

File: ffi/diplomat/tests/missing_apis.txt (hunk 1 above)
Comment (Manishearth) [reply ffi/diplomat/tests/missing_apis.txt:14]: I'd rather do specific names for now since the owners group is not the same group as the people doing FFI.

**Pull request merged**